File: python/plugins/relative_paths_plugin.py
from pathlib import Path
from typing import List, Dict, Union

from models.document import Document
from plugins.md2html_plugin import Md2HtmlPlugin
from utils import relativize_relative_resource_path

MODULE_DIR = Path(__file__).resolve().parent


class RelativePathsPlugin(Md2HtmlPlugin):

    def __init__(self):
        super().__init__()
        self.markers: List[str] = []
        self.paths: Dict[str, str] = {}
        self.relative_paths_handlers: List[RelativePathsMetadataHandler] = []

    def accept_data(self, data):
        self.assure_accept_data_once()
        self.validate_data_with_file(data, MODULE_DIR.joinpath('relative_paths_schema.json'))
        markers = data.get("markers")
        if markers and type(markers) == list:
            self.markers = markers
            self.paths = data["paths"]
        else:
            self.paths = data
        handler = RelativePathsMetadataHandler(self.paths)
        self.relative_paths_handlers = [(handler, m, False) for m in self.markers]

    def is_blank(self) -> bool:
        return not bool(self.paths)

    def page_metadata_handlers(self):
        return self.relative_paths_handlers

    def variables(self, doc: Document) -> dict:
        result = {}
        for k, v in self.paths.items():
            result[k] = relativize_relative_resource_path(v, doc.output_file)
        return result


class RelativePathsMetadataHandler:
    def __init__(self, paths: Dict[str, str]):
        self.paths = paths

    def accept_page_metadata(self, doc: Document, marker: str, metadata_str: str, metadata_section,
                             visited_markers: Union[Dict[str, None]] = None):

        path = self.paths.get(metadata_str.strip())
        if path is None:
            return metadata_section
        else:
            return relativize_relative_resource_path(path, doc.output_file)

File: java/src/main/java/world/md2html/plugins/RelativePathsPlugin.java
package world.md2html.plugins;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import world.md2html.options.argfile.ArgFileParseException;
import world.md2html.options.model.Document;
import world.md2html.utils.CheckedIllegalArgumentException;
import world.md2html.utils.UserError;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static world.md2html.utils.Utils.relativizeRelativePath;

public class RelativePathsPlugin extends AbstractMd2HtmlPlugin implements PageMetadataHandler {

    private Map<String, String> paths;
    private List<String> markers;
    private List<PageMetadataHandlerInfo> handlers;

    @Override
    public void acceptData(JsonNode data) throws ArgFileParseException {
        validateInputDataAgainstSchemaFromResource(data, "plugins/relative_paths_schema.json");

        JsonNode pathsNode;
        JsonNode markers = data.get("markers");
        if (markers instanceof ArrayNode) {
            List<String> markerList = new ArrayList<>();
            markers.forEach(node -> markerList.add(node.asText()));
            this.markers = markerList;
            pathsNode = data.get("paths");
        } else {
            this.markers = Collections.emptyList();
            pathsNode = data;
        }

        Map<String, String> pluginData = new HashMap<>();
        pathsNode.fields().forEachRemaining(entry -> pluginData.put(entry.getKey(),
                entry.getValue().asText()));
        this.paths = pluginData;

        this.handlers = this.markers.stream()
                .map(m -> new PageMetadataHandlerInfo(this, m, false))
                .collect(Collectors.toList());
    }

    @Override
    public boolean isBlank() {
        return this.paths.isEmpty();
    }

    @Override
    public List<PageMetadataHandlerInfo> pageMetadataHandlers() {
        return this.handlers;
    }

    @Override
    public Map<String, Object> variables(Document document) {
        Map<String, Object> variables = new HashMap<>();
        this.paths.forEach((k, v) -> {
            try {
                variables.put(k, relativizeRelativePath(v, document.getOutput()));
            } catch (CheckedIllegalArgumentException e) {
                throw new UserError("Error recalculating relative path '" + k + "'='" + v +
                        "' for page '" + document.getOutput() + "': " + e.getMessage());
            }
        });
        return variables;
    }

    @Override
    public String acceptPageMetadata(Document document, String marker, String metadata,
            String metadataSection) throws PageMetadataException {

        String path = this.paths.get(metadata.trim());
        if (path == null) {
            return metadataSection;
        } else {
            try {
                return relativizeRelativePath(path, document.getOutput());
            } catch (CheckedIllegalArgumentException e) {
                throw new PageMetadataException("Plugin '" + this.getClass().getSimpleName() +
                        "': Cannot relativize '" + path + "' against '" +
                        document.getOutput() + "': " + e.getMessage());
            }
        }
    }
}
